Several improvements to ARC shrinking

amotin opened the pull request:
### Motivation and Context
Since same time updating to Linux 6.6 kernel and increasing maximum ARC size in TrueNAS SCALE 24.04, we've started to receive multiple complains from people on excessive swapping, making systems unresponsive.  While I attribute significant part of the problem to the new Multi-Gen LRU code enabled in 6.6 kernel (disabling it helps), I ended up with this set of smaller tunings to ZFS side, trying to make it a bit nicer in this terrible environment.

### Description
 - When receiving memory pressure signal from OS be more strict trying to free some memory.  Otherwise kernel may come again and request much more.  Return as result how much arc_c was actually reduced due to this request, that may be less than requested.
 - On Linux when receiving direct reclaim from some file system (that may be ZFS) instead of ignoring request completely, just shrink the ARC, but do not wait for eviction.  Waiting there may cause deadlock.  Ignoring it as before may put extra pressure on other caches and/or swap, and cause OOM if nothing help.  While not waiting may result in more ARC evicted later, and may be too late if OOM killer activate right now, but I hope it to be better than doing nothing at all.
 - On Linux set arc_no_grow before waiting for reclaim, not after, or it may grow back while we are waiting.
 - On Linux add new parameter zfs_arc_shrinker_seeks to balance ARC eviction cost, relative to page cache and other subsystems.
 - Slightly update Linux arc_set_sys_free() math for new kernels.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

adamdmoss reviewed (commented):
Half of these changes are definitely :+1: (I've been running with similar local changes to track and return how much was actually evicted), the rest I feel neutral or suspicious about as commented.
FWIW have you tried zfs_arc_shrinker_limit=0 rather than the more-complicated approach of estimating eviction cost etc?  limit=0 allegedly used to cause arc collapse, but I've not been able to trigger than for a long time, at least in combination with eviction code that accounts for how much was actually evicted.

amotin commented on module/os/linux/zfs/arc_os.c:
```diff
@@ -325,13 +335,17 @@ arc_set_sys_free(uint64_t allmem)
 	 * Base wmark_low is 4 * the square root of Kbytes of RAM.
 	 */
-	long wmark = 4 * int_sqrt(allmem/1024) * 1024;
+	long wmark = int_sqrt(allmem / 1024 * 16) * 1024;
```
It is actually improving precision, even though not by much.  And primarily it is what newer kernels actually do.

amotin commented on module/os/linux/zfs/arc_os.c:
```diff
@@ -325,13 +335,17 @@ arc_set_sys_free(uint64_t allmem)
 	 */
 	wmark = MAX(wmark, 128 * 1024);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 7, 0)
+	wmark = MIN(wmark, 256 * 1024 * 1024);
```
All this chunk is a kernel-specific magic. I am just updating it for newer kernels.

amotin commented on module/zfs/arc.c:
```diff
@@ -4832,11 +4839,14 @@ arc_wait_for_eviction(uint64_t amount, boolean_t use_reserve)
-		if (!arc_evict_needed) {
-			arc_evict_needed = B_TRUE;
-			zthr_wakeup(arc_evict_zthr);
+		if (lax) {
```
When OS signals memory pressure, then something is not good and we should be more careful about overflows handling and should complete evictions before return.  If the same happen as part of routing I/O, there is not point to be strict.

amotin replied:
I've added more comments here.

amotin replied:
I was trying to find some word opposite of `aggressive` or `under_pressure` to not make it opposite in meaning to the following `use_reserve` argument.

amotin commented on man/man4/zfs.4:
```diff
@@ -831,6 +831,13 @@ even with a small average compressed block size of ~8 KiB.
+restore evicted page.
+Bigger values makes ARC more precious and evictions smaller, comparing to
+other kernel subsystems.
+Value of 4 means parity with page cache.
```
Considering this affects amount of eviction by order of magnitude, I can hardly believe why it should be hard-coded.  The other question is that thanks to zfs_arc_shrinker_limit we really for the most part ignore what kernel wants from us.  Partially it is because kernel at times is completely insane in its requests, and new mentioned Multi-Gen LRU brings it to extreme, sometimes requesting eviction quarter to half of ARC for no reason.  I've already sent complaint email to the author, hope it lead to something.

snajpa commented:
FWIW I think there's another source for excessive swapping - it might be caused by too high `zfs_abd_scatter_max_order`. In our setup, it only takes a few days until excessive reclaim kicks in, then we have to add a zram-based swap device. When we lower `zfs_abd_scatter_max_order` to below 3, the excessive reclaim doesn't of course disappear fully, as there are other sources of pressure in the kernel to get higher order buddies, but it is very noticeable (load drops by 100 on a machine with 600 1st level containers and tons more nested).

amotin commented:
@snajpa Yes, I was also thinking about `zfs_abd_scatter_max_order`.  I don't have own numbers, but my thinking was that on FreeBSD, where ARC allocates only individual PAGE_SIZE pages, it takes from OS the least convenient memory, while on Linux ARC always allocates the best contiguous chunks it can, that leaves other subsystems that are more sensitive to fragmentation to suffer.  Contiguous chunks should be good for I/O efficiency, and on FreeBSD I do measure some per-page overheads, but there must be some sweet spot.

snajpa commented:
@amotin I haven't looked at the code yet, but if it doesn't do it already, it might be worth allocating the memory with flags so it doesn't trigger _any_ reclaim at all and then decrement the requested order on fail

amotin commented:
> it might be worth allocating the memory with flags so it doesn't trigger _any_ reclaim at all and then decrement the requested order on fail

That is what ZFS does.  It tries to allocate big first, but if fails, requests smaller and smaller until get enough.  But that way it consumes all remaining big chunks first.

snajpa commented:
It actually seems to directly call `kvmalloc()` when `HAVE_KVMALLOC`. In the 6.8 source I'm looking at, `kvmalloc` seems to do [__GFP_NORETRY](https://github.com/torvalds/linux/blob/77a6e927e31cb96e7989eec57a64cdcf8bea82ab/mm/util.c#L611), for which the documentation says it does `one round of reclaim in this implementation`. I'm tempted to change that line to `kmalloc_flags &= ~__GFP_DIRECT_RECLAIM;` to see what happens :D not sure what to do (if anything) on ZFS level with this information though.

snajpa commented:
I've tried `bpftrace`ing `spl_kvmem` calls and it seems at least `dsl_dir_tempreserve_space` and `dmu_buf_hold_array_by_dnode` are calling `spl_kvmem` (which ends up with one round of reclaim). Running this on a comparatively pretty much idle staging node, yet it's IMHO way too many calls in too little time...

```
[root@node1.stg.vpsfree.cz]
 ~ # timeout --foreground 10 bpftrace -e 'kprobe:spl_kvmalloc{ printf("%s: %s(%d)\n", probe, comm, pid); }' | wc -l
231947
```

shodanshok commented:
`zfs_arc_shrinker_limit=10000` (default) seems to strongly favor ARC, forcing heavy swapping even when not needed. Adjusting `vm.swappiness` has only limited effect (unless setting it to `0`). 

Does this patch address this issue? Can the fix be implemented within the limit of `zfs_arc_shrinker_limit` and `zfs_arc_pc_percent`, without introducing yet another tunable? It is becoming quite difficult to tune a system to avoid excessive swap.

Side question: in general, why it is so difficult to "emulate" the behavior of linux pagecache in respect grow, reclaim and shrink?

Thanks.

amotin commented:
> `zfs_arc_shrinker_limit=10000` (default) seems to strongly favor ARC, forcing heavy swapping even when not needed. Adjusting `vm.swappiness` has only limited effect (unless setting it to `0`).

I do plan to set it to 0 in our TrueNAS builds, since we control kernel there, but I have no good ideas what to do about upstream, since some Linux kernels tend to request enormous eviction amounts, even though original motivation of its additions should no longer apply to most users.  The 10000 default IMHO is extremely low, if any value other than 0 there is correct at all.

> Does this patch address this issue? Can the fix be implemented within the limit of `zfs_arc_shrinker_limit`, `zfs_arc_shrink_shift` and `zfs_arc_pc_percent`, without introducing yet another tunable? It is becoming quite difficult to tune a system to avoid excessive swap.

It is not expected to fix the issue by itself, only polish some moments.  As I have told, at this point we have removed MGLRU from our kernels, that helped a lot with excessive swapping, and I am going to set `zfs_arc_shrinker_limit=0` and `zfs_arc_pc_percent=300` to make ARC adjust better.  The new tunable I've added is more for completeness, I do not insist on it and may remove if there are objections.

> Side question: in general, why it is so difficult to "emulate" the behavior of linux pagecache in respect grow, reclaim and shrink?

Because page cache does not use the crippled shrinker KPIs ZFS has to use.  All memory pressure in Linux is built around page cache, and everything else is secondary.  And the mentioned MGLRU brings it to extreme, that is why we had to disable it, but it is not a long-term solution.

robn commented on module/os/linux/zfs/arc_os.c:
```diff
@@ -72,11 +74,19 @@
+ * page.  Bigger values make ARC more precious and evictions smaller comparing
+ * to other kernel subsystems.  Value of 4 means parity with page cache.
+ */
+static int zfs_arc_shrinker_seeks = DEFAULT_SEEKS;
```
Where does `4` come from in the comment? I'm guessing it falls out of the math in `do_shrink_slab()`?

amotin replied:
Yes.  From my code reading it seems kernel applies memory pressure to page cache in a ` >> priority` fraction of size.  Same time `do_shrink_slab()` multiples it by 4 before passing it to `->scan_objects()`.

robn reviewed (approved):
Minor nits above.

I don't know much about the MGLRU (just read the overview) and haven't seen its effect for real, so I don't really have a great sense of what the problems are. But this change looks pretty light, makes sense, and the tuneable allows a little more adjustment as we learn more about it. I'm good with this.

shodanshok commented:
As this patch touches `zfs_arc_shrinker_limit`, any thoughts regarding https://github.com/openzfs/zfs/pull/16313#issuecomment-2198551588 ? Do you feel comfortable leaving `zfs_arc_shrinker_limit=10000`? The default value seems too small to me.

amotin commented:
> As this patch touches `zfs_arc_shrinker_limit`, any thoughts regarding [#16313 (comment)](https://github.com/openzfs/zfs/pull/16313#issuecomment-2198551588) ? Do you feel comfortable leaving `zfs_arc_shrinker_limit=10000`? The default value seems too small to me.

This patch actually does nothing about `zfs_arc_shrinker_limit` and for a reason.  While I don't like the current default, I don't see a good alternative.  If I would change it, I would change it to 0 and then try to kick Linux developers to be reasonable.  While ZFS uses anything other than 0, it means it does not follow kernel memory pressure requests, and in that situation I see hopeless to try and make kernel cooperate.

amotin commented:
I've decided once more reconsider `arc_is_overflowing()`.  Previously it made caller never wait for eviction of less than 1/512 of ARC size or `SPA_MAXBLOCKSIZE` (16MB), whatever is bigger.  But Linux starts reclaim process from 1/4096 (see `DEF_PRIORITY` of 12), which means first several iterations ZFS may not timely react on memory pressure, forcing more eviction from page cache and other caches, which may already be on minimum if most of memory is consumed by ARC.

The new code uses `zfs_max_recordsize` as a minimum wait threshold under memory pressure, which is still 16MB on 64bit platforms, but only 1MB on 32bit, which should be nicer to the last.  Not considering `zfs_arc_overflow_shift` under pressure allows to be more reactive on a large systems, where 1/512 of ARC may mean gigabytes of RAM, while kernel may need much less, but right now.

amotin commented:
BTW, some workarounds growing from my MGLRU complains: https://git.kernel.org/pub/scm/linux/kernel/git/akpm/25-new.git/tree/patches/mm-mglru-fix-overshooting-shrinker-memory.patch .

shodanshok commented:
> PS: Thinking more, with the current `zfs_arc_shrinker_limit` default of 10000 pages (that means only 40MB memory reclaim at a time under absolutely desperate pressure before OOM killer), I suppose ZFS could almost never react timely on the memory pressure on large systems. This should make it some better, while `zfs_arc_shrinker_limit` is still evil.

I agree. It should actually be ~160 MB (40 MB * 4 sublists), but the results does not change: under memory pressure, ARC force heavy swap and/or OOM. A more reasonable default for `zfs_arc_shrinker_limit` should be in the range of 128K pages, with no limit at all when direct reclaim is requested.

tonyhutter commented on module/os/linux/zfs/arc_os.c:
```diff
@@ -318,20 +316,26 @@ arc_set_sys_free(uint64_t allmem)
 	 */
 
 	/*
 	 * Base wmark_low is 4 * the square root of Kbytes of RAM.
 	 */
-	long wmark = 4 * int_sqrt(allmem/1024) * 1024;
+	long wmark = int_sqrt(allmem / 1024 * 16) * 1024;
```
Does this comment need to be updated with the new calculation?

amotin replied:
It is effectively the same within rounding error.  I can update if you insist, but I'd leave it as-is.

tonyhutter replied:
Ah I see, nevermind, leave as-is

tonyhutter commented:
I don't see any major issues.  Could you please rebase on master to get a good FreeBSD test run?
